File: python/pontificator.py
#!/usr/bin/env python3
import os
import sys
import subprocess

def convert_to_text(file_path):
    # Convert MDX file to plain text using Pandoc and custom filters
    subprocess.run([
        "python3", "python/clean_mdx.py", f"{file_path}.mdx"
    ])

def chunk_text(file_path):
    # Split the text into chunks
    subprocess.run([
        "python3",
        "python/split_into_chunks.py",
        f"{file_path}.txt"
    ])

    end_file_path = file_path.split('/')[-1]

    # Get list of chunk files starts with file_path-chunk- and is txt
    chunk_files = [
        file for file in os.listdir("example")
        if file.startswith(f"{end_file_path}-chunk-") and file.endswith(".txt")
    ]
    return chunk_files

def generate_audio(chunk_files):
    # Generate audio from chunked text
    for chunk_file in chunk_files:
        subprocess.run([
            "python3",
            "python/generate_audio.py",
            f"example/{chunk_file}"
        ])

def main():
    if len(sys.argv) != 2:
        print("Usage: python pontificator.py <file_path>")
        sys.exit(1)

    file_path = sys.argv[1]
    file_path_no_ext = os.path.splitext(file_path)[0]

    # Convert MDX file to plain text
    convert_to_text(file_path_no_ext)

    # Chunk the text into smaller files
    chunk_files = chunk_text(file_path_no_ext)

    # Generate audio from chunked text
    generate_audio(chunk_files)

    # Combine audio files into one
    subprocess.run([
        "python3",
        "python/combine_audio.py",
        file_path_no_ext
    ])

    # Remove chunk txt and mp3 files
    for chunk_file in chunk_files:
        mp3_file = chunk_file.replace('txt', 'mp3')
        os.remove(f"example/{chunk_file}")
        os.remove(f"example/{mp3_file}")

if __name__ == "__main__":
    main()

File: src/pontificator.ts
import * as fs from 'fs';
import path from 'path';
import { cleanMdx } from './cleanMdx';
import { splitIntoChunks } from './splitIntoChunks';
import { generateAudio } from './generateAudio';
import { combineAudio } from './combineAudio';

export const pontificator = async (): Promise<void> => {
    const filePath = process.argv[2];
    if (!filePath) {
        console.error('Usage: ts-node pontificator.ts <file_path>');
        process.exit(1);
    }

    // Resolve the provided path to an absolute path
    const absoluteFilePath = path.resolve(process.cwd(), filePath);

    // Extract the base name (filename without extension)
    const absoluteFilePathNoExt = absoluteFilePath.replace(/\.[^/.]+$/, '');

    // Extract the absolute path to containing directory
    const absoluteDirPath = path.dirname(absoluteFilePath);

    console.log('====================================================');
    console.log(`\x1b[36mStarting pontificator...\x1b[0m`);
    console.log('====================================================');
    console.log(`\x1b[32mFile to process:\x1b[0m       ${filePath}`);
    console.log(`\x1b[32mAbsolute path:\x1b[0m        ${absoluteFilePath}`);
    console.log('====================================================');

    // 1. Convert MDX file to plain text
    console.log(`\x1b[34mCleaning MDX file: ${absoluteFilePathNoExt}.mdx...\x1b[0m`);
    cleanMdx(`${absoluteFilePathNoExt}.mdx`);
    console.log(`\x1b[32mMDX file cleaned and saved as .txt\x1b[0m`);

    // 2. Split the text into smaller chunks
    console.log(`\x1b[34mSplitting ${absoluteFilePathNoExt}.txt into smaller chunks...\x1b[0m`);
    const chunkFilePaths = splitIntoChunks(`${absoluteFilePathNoExt}.txt`);
    console.log(`\x1b[32mCreated ${chunkFilePaths.length} chunk file(s):\x1b[0m`, chunkFilePaths);

    // 3. Generate audio from chunked text
    console.log('\x1b[34mGenerating audio for each chunk...\x1b[0m');
    for (const chunkFilePath of chunkFilePaths) {
        console.log(`  • Processing chunk file: ${chunkFilePath}`);
        await generateAudio(chunkFilePath);
        console.log(`  \x1b[32mAudio generated:\x1b[0m     ${chunkFilePath.replace('.txt', '.mp3')}`);
    }

    // 4. Combine audio files
    console.log('\x1b[34mCombining all chunked MP3 files...\x1b[0m');
    const success = await combineAudio(absoluteDirPath);
    if (!success) {
        console.error('\x1b[31mError combining audio files\x1b[0m');
        process.exit(1);
    }
    console.log('\x1b[32mAll chunked MP3 files combined!\x1b[0m');

    // 5. Clean up chunk files
    console.log('\x1b[34mCleaning up temporary chunk files...\x1b[0m');
    chunkFilePaths.forEach(chunkFile => {
        const mp3File = chunkFile.replace('.txt', '.mp3');
        fs.unlinkSync(chunkFile);
        fs.unlinkSync(mp3File);
        console.log(`  \x1b[31mDeleted:\x1b[0m ${chunkFile} and ${mp3File}`);
    });

    console.log('====================================================');
    console.log('\x1b[32mPontification complete!\x1b[0m');
    console.log('====================================================');
};

pontificator();
